### ISSUE TITLE
self.table hash contradiction

### ISSUE THREAD

**Issue by aldycool** [dpwe/audfprint#24]
Hi @dpwe,

I'm investigating on how the self.table in hash_table.py works when inserting and querying data. I found some confusion in this issue, please see my notes below:

```
# In hash_table.py, function get_hits():
# hashes = parameter query hash returned from wavfile2hashes (list of (time, hash) tuples)
# hash_ is taken straight from hashes, means the hash value is NOT masked
hash_ = hashes[ix][1]

# We check how many data contained in the self.counts for the given hash_
# hash_ is still NOT masked which is correct because self.counts stores unmasked hashes
nids = min(self.depth, self.counts[hash_])

# This is where it gets confusing, because we are trying to get the table values using the unmasked hash
tabvals = self.table[hash_, :nids]

# But the implementation of storing data into self.table in hash_table.py function store() is like this:
# sortedpairs = time-hash pairs ingested from sound file, returned from wavfile2hashes (list of (time, hash) tuples)
for time_, hash_ in sortedpairs:
    # This is still OK because we are getting the counts using the original hash (not masked)
    count = self.counts[hash_]
    time_ &= timemask
    # This is the point where the hash_ variable is replaced with the masked version of hash_
    hash_ &= hashmask
    val = (idval + time_) #.astype(np.uint32)
    if count < self.depth:
        # This is where we store into the self.table USING the HASHED VERSION of hash_, which is contradict the previous retrieval process of self.table
        self.table[hash_, count] = val

```

TL;DR; is, the way we store values into the hash table is using the masked hash as the key, and then we get the values from the hash table using the original / unmasked hash as the key. Can you kindly explain the situation please, thanks!

**Comment by dpwe** [comment]
Aldy -

Thanks for bringing this up.

The point of masking is to allow hash tables that are smaller than the
default 20 bit space: a smaller table can be constructed simply by taking
the lower N bits.

However, this feature hasn't been used by me, and likely isn't implemented
right.  The hashes formed from landmarks should always be ANDed with the
mask (which is always a string of N 1s in binary) to get the address.

self.counts should also use the masked hash values.

I'll fix this and add a test.

  DAn.

**Comment by dpwe** [comment]
Fixed in fd6cad58e37bb4937e83090576cf117a975172d1
